File: barelymusician/instruments/synth_instrument.cpp
#include "barelymusician/instruments/synth_instrument.h"

#include <array>
#include <cassert>

#include "barelymusician/barelymusician.h"
#include "barelymusician/dsp/oscillator.h"
#include "barelymusician/instruments/custom_instrument.h"

BarelyInstrumentDefinition BarelySynthInstrument_GetDefinition() {
  return barely::SynthInstrument::GetDefinition();
}

namespace barely {

namespace {

// Maximum number of voices allowed to be set.
constexpr int kMaxVoiceCount = 64;

}  // namespace

InstrumentDefinition SynthInstrument::GetDefinition() noexcept {
  static const std::array<ControlDefinition, static_cast<int>(Control::kCount)>
      control_definitions = {
          // Gain.
          ControlDefinition{Control::kGain, 1.0, 0.0, 1.0},
          // Oscillator type.
          ControlDefinition{Control::kOscillatorType, static_cast<double>(OscillatorType::kSine),
                            0.0, static_cast<double>(OscillatorType::kNoise)},
          // Attack.
          ControlDefinition{Control::kAttack, 0.05, 0.0, 60.0},
          // Decay.
          ControlDefinition{Control::kDecay, 0.0, 0.0, 60.0},
          // Sustain.
          ControlDefinition{Control::kSustain, 1.0, 0.0, 1.0},
          // Release.
          ControlDefinition{Control::kRelease, 0.25, 0.0, 60.0},
          // Number of voices.
          ControlDefinition{Control::kVoiceCount, 8, 1, kMaxVoiceCount},
      };
  return CustomInstrument::GetDefinition<SynthInstrument>(control_definitions, {});
}

// NOLINTNEXTLINE(bugprone-exception-escape)
SynthInstrument::SynthInstrument(int frame_rate) noexcept
    : voice_(SynthVoice(frame_rate), kMaxVoiceCount),
      gain_processor_(frame_rate),
      tuning_(GetStandardTuning()) {}

void SynthInstrument::Process(double* output_samples, int output_channel_count,
                              int output_frame_count) noexcept {
  for (int frame = 0; frame < output_frame_count; ++frame) {
    const double mono_sample = voice_.Next(0);
    for (int channel = 0; channel < output_channel_count; ++channel) {
      output_samples[output_channel_count * frame + channel] = mono_sample;
    }
  }
  gain_processor_.Process(output_samples, output_channel_count, output_frame_count);
}

// NOLINTNEXTLINE(bugprone-exception-escape)
void SynthInstrument::SetControl(int id, double value) noexcept {
  switch (static_cast<Control>(id)) {
    case Control::kGain:
      gain_processor_.SetGain(value);
      break;
    case Control::kOscillatorType:
      voice_.Update([value](SynthVoice* voice) noexcept {
        voice->generator().SetType(static_cast<OscillatorType>(static_cast<int>(value)));
      });
      break;
    case Control::kAttack:
      voice_.Update([value](SynthVoice* voice) noexcept { voice->envelope().SetAttack(value); });
      break;
    case Control::kDecay:
      voice_.Update([value](SynthVoice* voice) noexcept { voice->envelope().SetDecay(value); });
      break;
    case Control::kSustain:
      voice_.Update([value](SynthVoice* voice) noexcept { voice->envelope().SetSustain(value); });
      break;
    case Control::kRelease:
      voice_.Update([value](SynthVoice* voice) noexcept { voice->envelope().SetRelease(value); });
      break;
    case Control::kVoiceCount:
      voice_.Resize(static_cast<int>(value));
      break;
    default:
      assert(false);
      break;
  }
}

void SynthInstrument::SetNoteOff(int pitch) noexcept { voice_.Stop(pitch); }

void SynthInstrument::SetNoteOn(int pitch, double intensity) noexcept {
  voice_.Start(pitch, [frequency = tuning_.GetFrequency(pitch), intensity](SynthVoice* voice) {
    voice->generator().SetFrequency(frequency);
    voice->set_gain(intensity);
  });
}

void SynthInstrument::SetTuning(const TuningDefinition& tuning) noexcept { tuning_ = tuning; }

}  // namespace barely

File: platforms/unity/Assets/BarelyMusician/Scripts/Instruments/SynthInstrument.cs
using UnityEngine;

namespace Barely {
  /// Synthesizer oscillator type.
  public enum OscillatorType {
    /// Sine wave.
    [InspectorName("Sine")] SINE = 0,
    /// Sawtooth wave.
    [InspectorName("Saw")] SAW = 1,
    /// Square wave.
    [InspectorName("Square")] SQUARE = 2,
    /// White noise.
    [InspectorName("Noise")] NOISE = 3,
  }

  /// A representation of a simple polyphonic synthesizer instrument with an envelope.
  public class SynthInstrument : Instrument {
    /// Gain.
    [Range(0.0f, 1.0f)]
    public double Gain = 1.0;

    /// Oscillator type.
    public OscillatorType OscillatorType = OscillatorType.SINE;

    /// Envelope attack in seconds.
    [Range(0.0f, 60.0f)]
    public double Attack = 0.05;

    /// Envelope decay in seconds.
    [Range(0.0f, 60.0f)]
    public double Decay = 0.0;

    /// Envelope sustain.
    [Range(0.0f, 1.0f)]
    public double Sustain = 1.0;

    /// Envelope release in seconds.
    [Range(0.0f, 60.0f)]
    public double Release = 0.25;

    /// Number of voices.
    [Range(1, 64)]
    public int VoiceCount = 8;

    private void Update() {
      GetControl(0).Value = Gain;
      GetControl(1).Value = (double)OscillatorType;
      GetControl(2).Value = Attack;
      GetControl(3).Value = Decay;
      GetControl(4).Value = Sustain;
      GetControl(5).Value = Release;
      GetControl(6).Value = (double)VoiceCount;
    }
  }
}  // namespace Barely
